src/storage/store/column.cpp

```diff
@@ -227,96 +224,46 @@ void Column::scan(Transaction* transaction, const ChunkState& state, ColumnChunk
     auto cursor = PageUtils::getPageCursorForPos(startOffset, numValuesPerPage);
     cursor.pageIdx += state.metadata.pageIdx;
-    uint64_t numValuesScanned = 0u;
     KU_ASSERT((numValuesToScan + startOffset) <= state.metadata.numValues);
-    while (numValuesScanned < numValuesToScan) {
-        auto numValuesToReadInPage =
-            std::min(numValuesPerPage - cursor.elemPosInPage, numValuesToScan - numValuesScanned);
-        KU_ASSERT(isPageIdxValid(cursor.pageIdx, state.metadata));
-        readFromPage(transaction, cursor.pageIdx, [&](uint8_t* frame) -> void {
-            readToPageFunc(frame, cursor, columnChunk->getData(), numValuesScanned,
-                numValuesToReadInPage, state.metadata.compMeta);
-        });
-        numValuesScanned += numValuesToReadInPage;
-        cursor.nextPage();
-    }
-    KU_ASSERT(numValuesScanned == numValuesToScan);
+
+    const uint64_t numValuesScanned = columnReadWriter->readCompressedValuesToPage(transaction,
+        state, columnChunk->getData(), 0, startOffset, endOffset, readToPageFunc);
+
     columnChunk->setNumValues(numValuesScanned);
 }
 
 void Column::scan(Transaction* transaction, const ChunkState& state, offset_t startOffsetInGroup,
     offset_t endOffsetInGroup, uint8_t* result) {
-    auto cursor = getPageCursorForOffsetInGroup(startOffsetInGroup, state);
-    const auto numValuesToScan = endOffsetInGroup - startOffsetInGroup;
-    uint64_t numValuesScanned = 0;
-    while (numValuesScanned < numValuesToScan) {
-        uint64_t numValuesToScanInPage =
-            std::min(static_cast<uint64_t>(state.numValuesPerPage) - cursor.elemPosInPage,
-                numValuesToScan - numValuesScanned);
-        readFromPage(transaction, cursor.pageIdx, [&](uint8_t* frame) -> void {
-            readToPageFunc(frame, cursor, result, numValuesScanned, numValuesToScanInPage,
-                state.metadata.compMeta);
-        });
-        numValuesScanned += numValuesToScanInPage;
-        cursor.nextPage();
-    }
+    columnReadWriter->readCompressedValuesToPage(transaction, state, result, 0, startOffsetInGroup,
+        endOffsetInGroup, readToPageFunc);
 }
 
 void Column::scanInternal(Transaction* transaction, const ChunkState& state,
     offset_t startOffsetInChunk, row_idx_t numValuesToScan, ValueVector* nodeIDVector,
     ValueVector* resultVector) {
-    auto cursor = getPageCursorForOffsetInGroup(startOffsetInChunk, state);
     if (nodeIDVector->state->getSelVector().isUnfiltered()) {
-        scanUnfiltered(transaction, cursor, numValuesToScan, resultVector, state.metadata);
+        columnReadWriter->readCompressedValuesToVector(transaction, state, resultVector, 0,
+            startOffsetInChunk, startOffsetInChunk + numValuesToScan, readToVectorFunc);
     } else {
-        scanFiltered(transaction, cursor, numValuesToScan, nodeIDVector->state->getSelVector(),
-            resultVector, state.metadata);
-    }
-}
+        struct Filterer {
```

Comment:
Is this a pure refactoring? Is it necessary for this PR? I'm actually a bit lost why we add this abstraction. I think I need some more time to go through the changes related to `Column`. It seems there are some heavy refactoring going on, which I'm not sure if they should be done in this PR or not. Maybe let's discuss in person later.

Reply:
Discussed offline, keeping it this way (since this makes overriding the column writing behaviour easier)